otbox.py

```diff
@@ -25,42 +34,209 @@
+        """Flashes the given firmware file to the mote connected to the given serialport"""
+        pass
+
+    def on_mqtt_connect(self):
+        pass
+
+class OpenTestbed(Testbed):
+
+    INIT_PICTURE_URL              = 'https://upload.wikimedia.org/wikipedia/commons/7/74/Openwsn_logo.png'
+
+    def __init__(self, otbox):
+        super(OpenTestbed,self).__init__(otbox)
+
+        self.baudrate = 115200
+        self.mote_usb_devices = [
+            '/dev/openmote-b_1',
+            '/dev/openmote-b_2',
+            '/dev/openmote-b_3',
+            '/dev/openmote-b_4',
+        ]
+
+        self.firmware_eui64_retrieval = os.path.join(os.path.dirname(__file__), 'bootloaders', 'opentestbed',
+                                                     '01bsp_eui64_prog.ihex')
+        self.firmware_temp = os.path.join(os.path.dirname(__file__), 'bootloaders', 'opentestbed', 'firmware_mote.ihex')
+
+        # import image processing libs if opentestbed
+        from PIL import Image
+        from PIL import ImageFont
+        from PIL import ImageDraw
+
+    def bootload_mote(self, serialport, firmware_file):
+        return subprocess.Popen(
+                ['python', '../bootloaders/cc2538-bsl.py', '-e', '--bootloader-invert-lines', '-w', '-b', '400000',
+                 '-p', serialport, firmware_file], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    def _display_image(self):
+        import pygame
+        pygame.init()
+
+        size       = (pygame.display.Info().current_w, pygame.display.Info().current_h)
+        black      = 0, 0, 0
+        screen     = pygame.display.set_mode(size,pygame.FULLSCREEN)
+        while True:
+            if self.change_image_queue.empty()==False:
+                picture      = self.change_image_queue.get()
+                image        = pygame.image.fromstring(picture.tobytes(), picture.size, picture.mode)
+                imagerect    = image.get_rect()
+                screen.fill(black)
+                screen.blit(image, (240-picture.size[0]/2,160-picture.size[1]/2))
+                pygame.display.flip()
+                self.change_image_queue.task_done()
+            time.sleep(0.2)
+
+    def on_mqtt_connect(self):
+        # start heartbeat thread
+        currentThreads = _getThreadsName()
+
+        if IMAGE_THREAD_NAME not in currentThreads:
+            self.image_thread = threading.Thread(
+                name=IMAGE_THREAD_NAME,
+                target=self._display_image,
+            )
+            self.image_thread.start()
+            self._mqtt_handler_picturetoscreen('box','all',json.dumps({'token': 0, 'url': OpenTestbed.INIT_PICTURE_URL}))
+
+    def _mqtt_handler_picturetoscreen(self, deviceType, deviceId, payload):
+        '''
+        {{TESTBED}}/deviceType/box/deviceId/box1/cmd/picturetoscreen
+        '''
+        assert deviceType==DEVICETYPE_BOX
+        image = Image.open(requests.get(json.loads(payload)['url'], stream=True).raw)
```

Comment:
I got following error when testing on opentestbed:
```
Exception in thread mqtt_loop_thread:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 1481, in loop_forever
    rc = self.loop(timeout, max_packets)
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 1003, in loop
    rc = self.loop_read(max_packets)
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 1284, in loop_read
    rc = self._packet_read()
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 1849, in _packet_read
    rc = self._packet_handle()
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 2311, in _packet_handle
    return self._handle_connack()
  File "/usr/local/lib/python2.7/dist-packages/paho/mqtt/client.py", line 2372, in _handle_connack
    self.on_connect(self, self._userdata, flags_dict, result)
  File "otbox.py", line 305, in _on_mqtt_connect
    self.tb.on_mqtt_connect()
  File "otbox.py", line 119, in on_mqtt_connect
    self._mqtt_handler_picturetoscreen('box','all',json.dumps({'token': 0, 'url': OpenTestbed.INIT_PICTURE_URL}))
  File "otbox.py", line 126, in _mqtt_handler_picturetoscreen
    image = Image.open(requests.get(json.loads(payload)['url'], stream=True).raw)
NameError: global name 'Image' is not defined
```

Is that ok to put 
```
        from PIL import Image
        from PIL import ImageFont
        from PIL import ImageDraw
```
globally? is that too heavy to the machine in other testbed?

Reply:
I would avoid having a hard dependency on Pillow (PIL) for functionality which is specific to your deployments in Paris.

We can add these imports globally, but with a `try ... except ImportError` around it ?